internal/persistence/sql/relationtuples.go

```diff
@@ -18,105 +20,210 @@ import (
+		n, err := nm.GetNamespaceByName(ctx, st.Namespace)
+		if err != nil {
+			return err
+		}
+
+		r.SubjectID = sql.NullString{}
+		r.SubjectSetNamespaceID = sql.NullInt64{
+			Int64: n.ID,
+			Valid: true,
+		}
+		r.SubjectSetObject = sql.NullString{
+			String: st.Object,
+			Valid:  true,
+		}
+		r.SubjectSetRelation = sql.NullString{
+			String: st.Relation,
+			Valid:  true,
+		}
+	}
+	return nil
+}
+
+func (r *RelationTuple) fromInternal(ctx context.Context, p *Persister, rt *relationtuple.InternalRelationTuple) error {
+	nm, err := p.d.Config().NamespaceManager()
+	if err != nil {
+		return err
+	}
+
+	n, err := nm.GetNamespaceByName(ctx, rt.Namespace)
+	if err != nil {
+		return err
+	}
+
+	r.NetworkID = p.NetworkID(ctx)
+	r.NamespaceID = n.ID
+	r.Object = rt.Object
+	r.Relation = rt.Relation
+
+	return r.insertSubject(ctx, nm, rt.Subject)
 }
 
 func (p *Persister) insertRelationTuple(ctx context.Context, rel *relationtuple.InternalRelationTuple) error {
 	if rel.Subject == nil {
 		return errors.New("subject is not allowed to be nil")
 	}
 
-	n, err := p.namespaces.GetNamespace(ctx, rel.Namespace)
+	p.d.Logger().WithFields(rel.ToLoggerFields()).Trace("creating in database")
+
+	rt := &RelationTuple{
+		ID:         uuid.Must(uuid.NewV4()),
+		NetworkID:  p.NetworkID(ctx),
+		CommitTime: time.Now(),
+	}
+	if err := rt.fromInternal(ctx, p, rel); err != nil {
+		return err
+	}
+
+	if err := sqlcon.HandleError(
+		p.Connection(ctx).Create(rt),
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Persister) deleteRelationTupleSubjectID(ctx context.Context, r *relationtuple.InternalRelationTuple, s *relationtuple.SubjectID, n *namespace.Namespace, network uuid.UUID) error {
+	if err := p.Connection(ctx).RawQuery(
+		"DELETE FROM keto_relation_tuples WHERE nid = ? AND namespace_id = ? AND object = ? AND relation = ? AND subject_id = ?",
+		network,
+		n.ID,
+		r.Object,
+		r.Relation,
+		s.ID,
+	).Exec(); err != nil {
+		return sqlcon.HandleError(err)
+	}
+
+	return nil
+}
+
+func (p *Persister) deleteRelationTupleSubjectSet(ctx context.Context, r *relationtuple.InternalRelationTuple, s *relationtuple.SubjectSet, n *namespace.Namespace, network uuid.UUID) error {
+	nm, err := p.d.Config().NamespaceManager()
 	if err != nil {
 		return err
 	}
 
-	p.l.WithFields(rel.ToLoggerFields()).Trace("creating in database")
+	sn, err := nm.GetNamespaceByName(ctx, s.Namespace)
+	if err != nil {
+		return err
+	}
 
-	return sqlcon.HandleError(
-		p.connection(context.WithValue(ctx, namespaceContextKey, n)).Create(&relationTuple{
-			ID:         shardID,
-			Object:     rel.Object,
-			Relation:   rel.Relation,
-			Subject:    rel.Subject.String(),
-			CommitTime: time.Now(),
-		}),
-	)
+	if err := p.Connection(ctx).RawQuery(
+		"DELETE FROM keto_relation_tuples WHERE nid = ? AND namespace_id = ? AND object = ? AND relation = ? AND subject_set_namespace_id = ? AND subject_set_object = ? AND subject_set_relation = ?",
```

Comment:
I assume it is not possible to do this by the shard ID?

Reply:
The shard ID is completely internal and never exposed, so we cannot get it as part of a request.